### ISSUE TITLE
Model collapse issue

### ISSUE THREAD

**Issue by PapaMadeleine2022** [OML-Team/open-metric-learning#367]
I use `pipelines/features_extraction/extractor_cub/train_cub.yaml` to train model on my own dataset which 
 is normal.  I find that everytime I set `normalise_features: False`, the model will collapse, i.e., the model output almost same value when inputing two different images. Otherwise, when I set `normalise_features:True`, the model will not collapse.
Have you encountered this issue?

**Comment by AlekseySh** [comment]
@PapaMadeleine2022 Hi!

I had an issue when I set `normalise_features:True` and `margin: null`. 
It's also mentioned in the documentation of a pipeline.

Is that your case?

**Comment by PapaMadeleine2022** [comment]
> when

what's your issue? model collapse?

**Comment by PapaMadeleine2022** [comment]
> @PapaMadeleine2022 Hi!
> 
> I had an issue when I set `normalise_features:True` and `margin: null`. It's also mentioned in the documentation of a pipeline.
> 
> Is that your case?

When I set  `normalise_features:True` and `margin: null`, like:

```
extractor:
  name: vit
  args:
    normalise_features: True
    use_multi_scale: False
    weights: vits8_dino
    arch: vits8

criterion:
  name: triplet_with_miner
  args:
    need_logs: True
    margin: null
    reduction: mean
    miner:
      name: triplets_with_memory
      args:
        bank_size_in_batches: 50
        tri_expand_k: 2
```

The model will not collapse.

**Comment by AlekseySh** [comment]
Okay, I picked the wrong words. I meant that if you normalise features and the margin is null, the model's ability to converge is way lower (it may be seen from the underlying math). The training will end with a much lower metric but without Nans.

**Comment by AlekseySh** [comment]
I have a guess, that maybe the problem is not in feature normalisation. When I trained a model on the CARS dataset (which is similar to CUB in its size and structure), sometimes I saw Nans. Since it was a stochastic thing, it was hard to say which parameter is problematic (there were Nans on one seed, but not on the other). That time I realized that Nans disappear if I switch from hard mining to all mining. Check out if all mining also works for your case.

**Comment by AlekseySh** [comment]
You also can pass args from the config to Lightning Trainer, see: https://github.com/OML-Team/open-metric-learning/blob/main/tests/test_runs/test_pipelines/configs/train.yaml#L81. Thus, you can try to clip your gradients using the Lightnings' functionality.

**Comment by PapaMadeleine2022** [comment]
> seen

No `NaN` occurs in my training, but the trained model collapses.

**Comment by PapaMadeleine2022** [comment]
> @PapaMadeleine2022 Okay, you mean all the distances between embeddings are similar to zero? It's a known issue of TripletLoss. SoftTripletLoss (`margin: null`) is exact one of the ideas to fight this problem, described in this [paper](https://arxiv.org/abs/1703.07737). It explains why it helps you to avoid collapse.

As you suggest. I use  `SoftTripletLoss` and set `normalise_features: False` like:
```
extractor:
  name: vit
  args:
    normalise_features: False
    use_multi_scale: False
    weights: vits16_dino
    arch: vits16

criterion:
  name: triplet_with_miner
  args:
    need_logs: True  # set True to log values like positive distance, negative distance, number of active triplets and so on
    margin: null  # <null> means that we use the soft version of Triplet Loss here. In this case, features normalisation in the model has to be turned off so gradients can flow normally
    reduction: mean
    miner:
      name: all_triplets  # We collect all the possible triplets in the batch. Based on the sampler's parameters it's ~10^3 * 20^2 = 400.000 triplets
      args: {}

```

Although the model does not collapse. But the precision is very low. All the distances between embeddings are still a little number.

**Comment by AlekseySh** [comment]
Unfortunately, it may happen, there is not bullet point. Try to change more hyperparameters and check what is going on in your data.

**Comment by PapaMadeleine2022** [comment]
> there is not bullet point

Ok. Thanks.

**Comment by AlekseySh** [comment]
I meant "silver bullet", but I think you understood :)
